File: akkaStreamsModelServer/src/main/scala/com/lightbend/scala/akkastream/queryablestate/inmemory/RestServiceInMemory.scala
package com.lightbend.scala.akkastream.queryablestate.inmemory

import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.server.Directives.{complete, get, path}
import akka.http.scaladsl.server.Route
import akka.stream.ActorMaterializer

import scala.concurrent.ExecutionContextExecutor
// import akka.util.timeout  // See usage below.
import com.lightbend.scala.akkastream.modelserver.stage.ModelStateStore
import com.lightbend.scala.modelServer.model.ModelToServeStats
import de.heikoseeberger.akkahttpjackson.JacksonSupport

/**
 * Implements Queryable State for the Akka Streams-based model scoring application.
 * Uses Akka HTTP to implement this capability, with state held in memory.
 * Note that a production implementation might need better scalability, if it's used
 * heavily and also we ignore security considerations here!
 */
object RestServiceInMemory {

  // Serve model status: http://localhost:5500/state
  def startRest(service: ModelStateStore)(implicit system: ActorSystem, materializer: ActorMaterializer): Unit = {

    implicit val executionContext: ExecutionContextExecutor = system.dispatcher
    // Use with HTTP methods that accept an implicit timeout argument
    // implicit val timeout = Timeout(10.seconds)
    val host = "127.0.0.1"
    val port = 5500
    val routes = QueriesAkkaHttpResource.storeRoutes(service)

    Http().bindAndHandle(routes, host, port) map
      { binding => println(s"Starting models observer on port ${binding.localAddress}") } recover {
      case ex =>
        println(s"Models observer could not bind to $host:$port - ${ex.getMessage}")
    }
  }
}

object QueriesAkkaHttpResource extends JacksonSupport {

  def storeRoutes(service: ModelStateStore): Route =
    get {
      path("state") {
        val info: ModelToServeStats = service.getCurrentServingInfo
        complete(info)
      }
    }
}

File: akkaStreamsModelServer/src/main/java/com/lightbend/java/akkastreams/queryablestate/inmemory/RestServiceInMemory.java
package com.lightbend.java.akkastreams.queryablestate.inmemory;

import akka.NotUsed;
import akka.actor.ActorSystem;
import akka.http.javadsl.ConnectHttp;
import akka.http.javadsl.Http;
import akka.http.javadsl.ServerBinding;
import akka.http.javadsl.marshallers.jackson.Jackson;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.model.HttpResponse;
import akka.http.javadsl.server.AllDirectives;
import akka.http.javadsl.server.Route;
import akka.stream.ActorMaterializer;
import akka.stream.javadsl.Flow;
import akka.util.Timeout;
import com.lightbend.java.akkastreams.modelServer.stage.ReadableModelStore;

import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;

public class RestServiceInMemory extends AllDirectives {

    private static final String host = "localhost";
    private static final int port = 5500;
    private static final Timeout askTimeout = Timeout.apply(5, TimeUnit.SECONDS);

    private ReadableModelStore reader = null;

    // Serve model status: http://localhost:5500/state
    public static void startRest(ActorSystem system, ActorMaterializer materializer, ReadableModelStore reader) {
        RestServiceInMemory resource = new RestServiceInMemory(reader);
        Flow<HttpRequest, HttpResponse, NotUsed> routeFlow = resource.createRoute().flow(system, materializer);
        Http http = Http.get(system);
        CompletionStage<ServerBinding> binding = http.bindAndHandle(routeFlow,
                ConnectHttp.toHost(host,port), materializer);
        System.out.println("Starting models observer on host " + host + " port " + port);
    }

    public RestServiceInMemory(ReadableModelStore reader){ this.reader = reader; }

    public Route createRoute() {
        return route(
                path("state", () ->
                        get(() -> completeOK(reader.getCurrentServingInfo(), Jackson.marshaller())))
        );
    }

}
